include/convert.h

```diff
@@ -12,88 +12,111 @@
 #include <target_device.h>
 #include <register_traits.h>
 #include <math_helper.cuh>
+#include <constexpr_for.h>
 
 namespace quda
 {
 
+#ifdef _NVHPC_CUDA
+#pragma diag_suppress no_device_stack
+#endif
+
   /**
    * Traits for determining the maximum and inverse maximum
    * value of a (signed) char and short. Relevant for
    * fixed-precision types.
    */
-  template <typename T> struct fixedMaxValue {
+  template <typename T1> struct fixedMaxValue {
     static constexpr float value = 0.0f;
   };
   template <> struct fixedMaxValue<short> {
     static constexpr float value = 32767.0f;
   };
-  template <> struct fixedMaxValue<short2> {
-    static constexpr float value = 32767.0f;
-  };
-  template <> struct fixedMaxValue<short4> {
-    static constexpr float value = 32767.0f;
-  };
-  template <> struct fixedMaxValue<short8> {
-    static constexpr float value = 32767.0f;
-  };
+
   template <> struct fixedMaxValue<int8_t> {
     static constexpr float value = 127.0f;
   };
-  template <> struct fixedMaxValue<char2> {
-    static constexpr float value = 127.0f;
-  };
-  template <> struct fixedMaxValue<char4> {
-    static constexpr float value = 127.0f;
-  };
-  template <> struct fixedMaxValue<char8> {
-    static constexpr float value = 127.0f;
-  };
 
-  template <typename T> struct fixedInvMaxValue {
+  template <typename T1> struct fixedInvMaxValue {
     static constexpr float value = 3.402823e+38f;
   };
+
   template <> struct fixedInvMaxValue<short> {
     static constexpr float value = 3.0518509476e-5f;
   };
-  template <> struct fixedInvMaxValue<short2> {
-    static constexpr float value = 3.0518509476e-5f;
-  };
-  template <> struct fixedInvMaxValue<short4> {
-    static constexpr float value = 3.0518509476e-5f;
-  };
-  template <> struct fixedInvMaxValue<short8> {
-    static constexpr float value = 3.0518509476e-5f;
-  };
+
   template <> struct fixedInvMaxValue<int8_t> {
     static constexpr float value = 7.874015748031e-3f;
   };
-  template <> struct fixedInvMaxValue<char2> {
-    static constexpr float value = 7.874015748031e-3f;
-  };
-  template <> struct fixedInvMaxValue<char4> {
-    static constexpr float value = 7.874015748031e-3f;
-  };
-  template <> struct fixedInvMaxValue<char8> {
-    static constexpr float value = 7.874015748031e-3f;
+
+  /**
+     @brief Regular integer to float used on the host
+  */
+  template <bool is_device> struct i2f {
+    template <typename T> constexpr float operator()(int a, T) { return static_cast<float>(a); }
+    template <typename T> constexpr float2 operator()(int a, int b, T)
+    {
+      return {static_cast<float>(a), static_cast<float>(b)};
+    }
   };
 
-  template <typename T> constexpr float i2f(T a)
-  {
-#ifndef QUDA_ALTERNATIVE_I_TO_F
-    return static_cast<float>(a);
-#else
-    // will work for up to 23-bit int
-    int32_t i = a + 0x4B400000;
-    float &f = reinterpret_cast<float &>(i);
-    return f - 12582912.0f;
+#if QUDA_ALTERNATIVE_I_TO_F == 100
```

Comment:
Add a smidge of documentation to explain what's going on either here or the first place `i2f_i` is used (a few functions down)

Reply:
https://github.com/lattice/quda/commit/d247829a5cae3d13d5fdc747fc46d77790520673